api/lib/opentelemetry/trace/span_context.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Trace
+    # A SpanContext contains the state that must propagate to child @see Spans and across process boundaries.
+    # It contains the identifiers (a @see TraceId and @see SpanId) associated with the @see Span and a set of
+    # @see TraceOptions.
+    #
+    # Design note: MRI optimizes storage of objects with 3 or fewer member variables, inlining the fields into the
+    # RVALUE. To take advantage of this, we avoid initializing @tracestate if undefined.
```

Comment:
These design notes are valuable, but I'd put them inline in the code where they won't get included in the {r,yar}doc since users shouldn't care.